README
Projektor
=========

Projektor is a document viewer runnning on DirectFB.


NEWS
0.9.1
-----
- Meson build system
- Follow MuPDF API changes
- Add DjVu and Poppler support


src/mupdf.c
/*
   This file is part of Projektor.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with this program; if not, write to the Free Software Foundation, Inc.,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
*/

#include "documentprovider.h"
#include <direct/memcpy.h>
#include <fitz.h>

extern DirectLink *documentproviders;

/**********************************************************************************************************************/

typedef struct {
     IDirectFB           *idirectfb;

     fz_context          *ctx;
     fz_document         *doc;

     DocumentDescription  desc;
} DocumentProvider_MuPDF_data;

/**********************************************************************************************************************/

static DFBResult
DocumentProvider_MuPDF_Init( DocumentProvider *thiz,
                             const char       *filename,
                             IDirectFB        *idirectfb )
{
     DFBResult                    ret = DFB_FAILURE;
     DocumentProvider_MuPDF_data *data;

     data = D_CALLOC( 1, sizeof(DocumentProvider_MuPDF_data) );
     if (!data)
          return D_OOM();;

     data->idirectfb = idirectfb;

     data->ctx = fz_new_context( NULL, NULL, FZ_STORE_DEFAULT );
     if (!data->ctx)
          goto error;

     fz_try( data->ctx ) {
          data->doc = fz_open_document( data->ctx, filename );
     }
     fz_catch( data->ctx ) {
          goto error;
     }

     if (strrchr( filename, '/' ))
          snprintf( data->desc.title, DOCUMENT_DESC_TITLE_LENGTH, strrchr( filename, '/' ) + 1 );
     else
          snprintf( data->desc.title, DOCUMENT_DESC_TITLE_LENGTH, filename );

     data->desc.num_pages = fz_count_pages( data->doc );

     thiz->priv = data;

     return DFB_OK;

error:
     if (data->ctx)
          fz_free_context( data->ctx );

     D_FREE( data );

     return ret;
}

static DFBResult
DocumentProvider_MuPDF_Term( DocumentProvider *thiz )
{
     DocumentProvider_MuPDF_data *data = thiz->priv;

     fz_close_document( data->doc );

     fz_free_context( data->ctx );

     D_FREE( data );

     return DFB_OK;
}

static DFBResult
DocumentProvider_MuPDF_GetDescription( DocumentProvider    *thiz,
                                       DocumentDescription *ret_desc )
{
     DocumentProvider_MuPDF_data *data = thiz->priv;

     if (!ret_desc)
          return DFB_INVARG;

     *ret_desc = data->desc;

     return DFB_OK;
}

static DFBResult
DocumentProvider_MuPDF_RenderPage( DocumentProvider  *thiz,
                                   int                pageno,
                                   float              zoom,
                                   IDirectFBSurface **ret_surface )
{
     DFBResult                    ret = DFB_FAILURE;
     DFBSurfaceDescription        desc;
     int                          pitch;
     int                          y;
     fz_matrix                    matrix;
     fz_rect                      rect;
     fz_irect                     irect;
     void                        *ptr;
     unsigned char               *src;
     IDirectFBSurface            *surface;
     fz_device                   *device = NULL;
     fz_page                     *page   = NULL;
     fz_pixmap                   *pixmap = NULL;
     DocumentProvider_MuPDF_data *data   = thiz->priv;

     fz_try( data->ctx ) {
          page = fz_load_page( data->doc, pageno - 1 );
     }
     fz_catch( data->ctx ) {
          goto out;
     }

     fz_rotate( &matrix, 0 );
     fz_pre_scale( &matrix, zoom, zoom );
     fz_bound_page( data->doc, page, &rect );
     fz_transform_rect( &rect, &matrix );
     fz_round_rect( &irect, &rect );

     desc.flags       = DSDESC_WIDTH | DSDESC_HEIGHT | DSDESC_PIXELFORMAT;
     desc.width       = irect.x1 - irect.x0;
     desc.height      = irect.y1 - irect.y0;
     desc.pixelformat = DSPF_ABGR;

     fz_try( data->ctx ) {
          pixmap = fz_new_pixmap_with_bbox( data->ctx, fz_device_rgb, &irect );

          device = fz_new_draw_device( data->ctx, pixmap );

          fz_run_page( data->doc, page, device, &matrix, NULL );
     }
     fz_catch( data->ctx ) {
          goto out;
     }

     ret = data->idirectfb->CreateSurface( data->idirectfb, &desc, &surface );
     if (ret)
          goto out;

     surface->Lock( surface, DSLF_WRITE, &ptr, &pitch );

     src = fz_pixmap_samples( data->ctx, pixmap );

     for (y = 0; y < desc.height; y++) {
          direct_memcpy( ptr, src, desc.width * 4 );

          src += desc.width * 4;
          ptr += pitch;
     }

     surface->Unlock( surface );

     *ret_surface = surface;

out:
     if (device)
          fz_free_device( device );

     if (pixmap)
          fz_drop_pixmap( data->ctx, pixmap );

     if (page)
          fz_free_page( data->doc, page );

     return ret;
}

static DocumentProvider mupdf_provider = {
     .impl           = "MuPDF",
     .Init           = DocumentProvider_MuPDF_Init,
     .Term           = DocumentProvider_MuPDF_Term,
     .GetDescription = DocumentProvider_MuPDF_GetDescription,
     .RenderPage     = DocumentProvider_MuPDF_RenderPage,
};

__attribute__((constructor))
static void
DocumentProvider_MuPDF_ctor()
{
     direct_list_append( &documentproviders, &mupdf_provider.link );
}


src/poppler.c
/*
   This file is part of Projektor.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with this program; if not, write to the Free Software Foundation, Inc.,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
*/

#include "documentprovider.h"
#include <direct/memcpy.h>
#include <poppler.h>

extern DirectLink *documentproviders;

/**********************************************************************************************************************/

typedef struct {
     IDirectFB           *idirectfb;

     PopplerDocument     *doc;

     DocumentDescription  desc;
} DocumentProvider_Poppler_data;

/**********************************************************************************************************************/

static DFBResult
DocumentProvider_Poppler_Init( DocumentProvider *thiz,
                               const char       *filename,
                               IDirectFB        *idirectfb )
{
     DFBResult                      ret = DFB_FAILURE;
     char                           uri[PATH_MAX];
     DocumentProvider_Poppler_data *data;

     data = D_CALLOC( 1, sizeof(DocumentProvider_Poppler_data) );
     if (!data)
          return D_OOM();;

     data->idirectfb = idirectfb;

     if (g_strstr_len( filename, -1, "://" )) {
          g_strlcpy( uri, filename, sizeof(uri) );
     }
     else {
          if (*filename == '/')
               g_snprintf( uri, sizeof(uri), "file://%s", filename );
          else
               g_snprintf( uri, sizeof(uri), "file://%s/%s", g_get_current_dir(), filename );
     }

     data->doc = poppler_document_new_from_file( uri, NULL, NULL );
     if (!data->doc)
          goto error;

     if (strrchr( filename, '/' ))
          snprintf( data->desc.title, DOCUMENT_DESC_TITLE_LENGTH, strrchr( filename, '/' ) + 1 );
     else
          snprintf( data->desc.title, DOCUMENT_DESC_TITLE_LENGTH, filename );

     data->desc.num_pages = poppler_document_get_n_pages( data->doc );

     thiz->priv = data;

     return DFB_OK;

error:
     D_FREE( data );

     return ret;
}

static DFBResult
DocumentProvider_Poppler_Term( DocumentProvider *thiz )
{
     DocumentProvider_Poppler_data *data = thiz->priv;

     g_object_unref( data->doc );

     D_FREE( data );

     return DFB_OK;
}

static DFBResult
DocumentProvider_Poppler_GetDescription( DocumentProvider    *thiz,
                                         DocumentDescription *ret_desc )
{
     DocumentProvider_Poppler_data *data = thiz->priv;

     if (!ret_desc)
          return DFB_INVARG;

     *ret_desc = data->desc;

     return DFB_OK;
}

static DFBResult
DocumentProvider_Poppler_RenderPage( DocumentProvider  *thiz,
                                     int                pageno,
                                     float              zoom,
                                     IDirectFBSurface **ret_surface )
{
     DFBResult                      ret = DFB_FAILURE;
     DFBSurfaceDescription          desc;
     int                            pitch;
     int                            y;
     double                         width;
     double                         height;
     cairo_status_t                 status;
     void                          *ptr;
     unsigned char                 *src;
     IDirectFBSurface              *surface;
     cairo_t                       *cairo  = NULL;
     PopplerPage                   *page   = NULL;
     cairo_surface_t               *pixmap = NULL;
     DocumentProvider_Poppler_data *data   = thiz->priv;

     page = poppler_document_get_page( data->doc, pageno - 1 );
     if (!page)
          goto out;

     poppler_page_get_size( page, &width, &height );

     desc.flags       = DSDESC_WIDTH | DSDESC_HEIGHT | DSDESC_PIXELFORMAT;
     desc.width       = width  * zoom + 0.5f;
     desc.height      = height * zoom + 0.5f;
     desc.pixelformat = DSPF_ARGB;

     pixmap = cairo_image_surface_create( CAIRO_FORMAT_ARGB32, desc.width, desc.height );
     status = cairo_surface_status( pixmap );
     if (status)
          goto out;

     cairo  = cairo_create( pixmap );
     status = cairo_status( cairo );
     if (status)
          goto out;

     cairo_scale( cairo, zoom, zoom );

     poppler_page_render( page, cairo );

     ret = data->idirectfb->CreateSurface( data->idirectfb, &desc, &surface );
     if (ret)
          goto out;

     surface->Lock( surface, DSLF_WRITE, &ptr, &pitch );

     src = cairo_image_surface_get_data( pixmap );

     for (y = 0; y < desc.height; y++) {
          direct_memcpy( ptr, src, desc.width * 4 );

          src += desc.width * 4;
          ptr += pitch;
     }

     surface->Unlock( surface );

     *ret_surface = surface;

out:
     if (cairo)
          cairo_destroy( cairo );

     if (pixmap)
          cairo_surface_destroy( pixmap );

     if (page)
          g_object_unref( page );

     return ret;
}

static DocumentProvider poppler_provider = {
     .impl           = "Poppler",
     .Init           = DocumentProvider_Poppler_Init,
     .Term           = DocumentProvider_Poppler_Term,
     .GetDescription = DocumentProvider_Poppler_GetDescription,
     .RenderPage     = DocumentProvider_Poppler_RenderPage,
};

__attribute__((constructor))
static void
DocumentProvider_Poppler_ctor()
{
     direct_list_append( &documentproviders, &poppler_provider.link );
}


src/documentprovider.h
/*
   This file is part of Projektor.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with this program; if not, write to the Free Software Foundation, Inc.,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
*/

#include <direct/list.h>
#include <directfb.h>

/*
 * Information about a document.
 */

#define DOCUMENT_DESC_TITLE_LENGTH 255

typedef struct {
     char title[DOCUMENT_DESC_TITLE_LENGTH];
     int  num_pages;
} DocumentDescription;

/*
 * Document provider interface.
 */

typedef struct _DocumentProvider DocumentProvider;

struct _DocumentProvider {
     DirectLink   link;
     const char  *impl;
     void        *priv;

     DFBResult  (*Init)          ( DocumentProvider *thiz, const char *filename, IDirectFB *idirectfb );
     DFBResult  (*Term)          ( DocumentProvider *thiz );
     DFBResult  (*GetDescription)( DocumentProvider *thiz, DocumentDescription *ret_desc );
     DFBResult  (*RenderPage)    ( DocumentProvider *thiz, int pageno, float zoom, IDirectFBSurface **ret_surface );
};


src/djvu.c
/*
   This file is part of Projektor.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with this program; if not, write to the Free Software Foundation, Inc.,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
*/

#include "documentprovider.h"
#include <direct/memcpy.h>
#include <libdjvu/ddjvuapi.h>

extern DirectLink *documentproviders;

/**********************************************************************************************************************/

typedef struct {
     IDirectFB           *idirectfb;

     ddjvu_context_t     *ctx;
     ddjvu_document_t    *doc;

     DocumentDescription  desc;
} DocumentProvider_DjVu_data;

/**********************************************************************************************************************/

static DFBResult
DocumentProvider_DjVu_Init( DocumentProvider *thiz,
                            const char       *filename,
                            IDirectFB        *idirectfb )
{
     DFBResult                    ret = DFB_FAILURE;
     DocumentProvider_DjVu_data *data;

     data = D_CALLOC( 1, sizeof(DocumentProvider_DjVu_data) );
     if (!data)
          return D_OOM();;

     data->idirectfb = idirectfb;

     data->ctx = ddjvu_context_create( "DjVu" );
     if (!data->ctx)
          goto error;

     data->doc = ddjvu_document_create_by_filename( data->ctx, filename, TRUE );
     if (!data->ctx)
          goto error;

     ddjvu_message_wait( data->ctx );
     ddjvu_message_pop( data->ctx );

     if (strrchr( filename, '/' ))
          snprintf( data->desc.title, DOCUMENT_DESC_TITLE_LENGTH, strrchr( filename, '/' ) + 1 );
     else
          snprintf( data->desc.title, DOCUMENT_DESC_TITLE_LENGTH, filename );

     data->desc.num_pages = ddjvu_document_get_pagenum( data->doc );

     thiz->priv = data;

     return DFB_OK;

error:
     if (data->ctx)
          ddjvu_context_release( data->ctx );

     D_FREE( data );

     return ret;
}

static DFBResult
DocumentProvider_DjVu_Term( DocumentProvider *thiz )
{
     DocumentProvider_DjVu_data *data = thiz->priv;

     ddjvu_document_release( data->doc );

     ddjvu_context_release( data->ctx );

     D_FREE( data );

     return DFB_OK;
}

static DFBResult
DocumentProvider_DjVu_GetDescription( DocumentProvider    *thiz,
                                      DocumentDescription *ret_desc )
{
     DocumentProvider_DjVu_data *data = thiz->priv;

     if (!ret_desc)
          return DFB_INVARG;

     *ret_desc = data->desc;

     return DFB_OK;
}

static DFBResult
DocumentProvider_DjVu_RenderPage( DocumentProvider  *thiz,
                                  int                pageno,
                                  float              zoom,
                                  IDirectFBSurface **ret_surface )
{
     DFBResult                   ret = DFB_FAILURE;
     DFBSurfaceDescription       desc;
     int                         pitch;
     int                         y;
     int                         dpi;
     ddjvu_rect_t                rect;
     void                       *ptr;
     unsigned char              *src;
     IDirectFBSurface           *surface;
     ddjvu_format_t             *format = NULL;
     ddjvu_page_t               *page   = NULL;
     char                       *pixmap = NULL;
     DocumentProvider_DjVu_data *data   = thiz->priv;

     page = ddjvu_page_create_by_pageno( data->doc, pageno - 1 );
     if (!page)
          goto out;

     while (!ddjvu_page_decoding_done( page )) {
          ddjvu_message_wait( data->ctx );
          ddjvu_message_pop( data->ctx );
     }

     dpi = ddjvu_page_get_resolution( page );

     desc.flags       = DSDESC_WIDTH | DSDESC_HEIGHT | DSDESC_PIXELFORMAT;
     desc.width       = ddjvu_page_get_width( page )  * 100 * zoom / dpi;
     desc.height      = ddjvu_page_get_height( page ) * 100 * zoom / dpi;
     desc.pixelformat = DSPF_RGB24;

     pixmap = D_MALLOC( desc.width * desc.height * 4 );
     if (!pixmap)
          goto out;

     format = ddjvu_format_create( DDJVU_FORMAT_BGR24, 0, NULL );
     if (!format)
          goto out;

     ddjvu_format_set_row_order( format, 1 );

     rect.x = 0;
     rect.y = 0;
     rect.w = desc.width;
     rect.h = desc.height;

     ddjvu_page_render( page, DDJVU_RENDER_COLOR, &rect, &rect, format, desc.width * 3, pixmap );

     ret = data->idirectfb->CreateSurface( data->idirectfb, &desc, &surface );
     if (ret)
          goto out;

     surface->Lock( surface, DSLF_WRITE, &ptr, &pitch );

     src = (unsigned char*) pixmap;

     for (y = 0; y < desc.height; y++) {
          direct_memcpy( ptr, src, desc.width * 3 );

          src += desc.width * 3;
          ptr += pitch;
     }

     surface->Unlock( surface );

     *ret_surface = surface;

out:
     if (format)
          ddjvu_format_release( format );

     if (pixmap)
          D_FREE( pixmap );

     if (page)
          ddjvu_page_release( page );

     return ret;
}

static DocumentProvider djvu_provider = {
     .impl           = "DjVu",
     .Init           = DocumentProvider_DjVu_Init,
     .Term           = DocumentProvider_DjVu_Term,
     .GetDescription = DocumentProvider_DjVu_GetDescription,
     .RenderPage     = DocumentProvider_DjVu_RenderPage,
};

__attribute__((constructor))
static void
DocumentProvider_DjVu_ctor()
{
     direct_list_append( &documentproviders, &djvu_provider.link );
}
